File: app/src/main/java/guepardoapps/mycoins/basic/utils/Logger.java
package guepardoapps.mycoins.basic.utils;

import android.support.annotation.NonNull;
import android.util.Log;

import java.io.Serializable;

public class Logger implements Serializable {

    private static final long serialVersionUID = -6278387904140268519L;

    private String _tag;
    private boolean _debuggingEnabled;

    public Logger(@NonNull String tag, boolean debuggingEnabled) {
        _tag = tag;
        _debuggingEnabled = debuggingEnabled;
        Debug("Created logger for " + _tag + "! Is Enabled: " + String.valueOf(_debuggingEnabled));
    }

    public Logger(@NonNull String tag) {
        this(tag, true);
    }

    public void SetEnable(boolean debuggingEnabled) {
        _debuggingEnabled = debuggingEnabled;
    }

    public boolean GetEnable() {
        return _debuggingEnabled;
    }

    public void Verbose(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.v(_tag, message);
            }
        }
    }

    public void Verbose(int message) {
        if (_debuggingEnabled) {
            Log.v(_tag, String.valueOf(message));
        }
    }

    public void Verbose(double message) {
        if (_debuggingEnabled) {
            Log.v(_tag, String.valueOf(message));
        }
    }

    public void Verbose(boolean message) {
        if (_debuggingEnabled) {
            Log.v(_tag, String.valueOf(message));
        }
    }

    public void Debug(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.d(_tag, message);
            }
        }
    }

    public void Debug(int message) {
        if (_debuggingEnabled) {
            Log.d(_tag, String.valueOf(message));
        }
    }

    public void Debug(double message) {
        if (_debuggingEnabled) {
            Log.d(_tag, String.valueOf(message));
        }
    }

    public void Debug(boolean message) {
        if (_debuggingEnabled) {
            Log.d(_tag, String.valueOf(message));
        }
    }

    public void Information(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.i(_tag, message);
            }
        }
    }

    public void Information(int message) {
        if (_debuggingEnabled) {
            Log.i(_tag, String.valueOf(message));
        }
    }

    public void Information(double message) {
        if (_debuggingEnabled) {
            Log.i(_tag, String.valueOf(message));
        }
    }

    public void Information(boolean message) {
        if (_debuggingEnabled) {
            Log.i(_tag, String.valueOf(message));
        }
    }

    public void Warning(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.w(_tag, message);
            }
        }
    }

    public void Warning(int message) {
        if (_debuggingEnabled) {
            Log.w(_tag, String.valueOf(message));
        }
    }

    public void Warning(double message) {
        if (_debuggingEnabled) {
            Log.w(_tag, String.valueOf(message));
        }
    }

    public void Warning(boolean message) {
        if (_debuggingEnabled) {
            Log.w(_tag, String.valueOf(message));
        }
    }

    public void Error(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.e(_tag, message);
            }
        }
    }

    public void Error(int message) {
        if (_debuggingEnabled) {
            Log.e(_tag, String.valueOf(message));
        }
    }

    public void Error(double message) {
        if (_debuggingEnabled) {
            Log.e(_tag, String.valueOf(message));
        }
    }

    public void Error(boolean message) {
        if (_debuggingEnabled) {
            Log.e(_tag, String.valueOf(message));
        }
    }
}

File: app/src/main/java/guepardoapps/mycoins/utils/Logger.kt
package guepardoapps.mycoins.utils

import android.content.Context
import android.support.annotation.NonNull
import android.util.Log
import guepardoapps.mycoins.database.logging.DbLogging
import guepardoapps.mycoins.database.logging.DbLog
import guepardoapps.mycoins.database.logging.Severity
import java.sql.Date
import java.util.*

internal class Logger private constructor() {
    var loggingEnabled: Boolean = true
    private var writeToDatabaseEnabled: Boolean = true
    private var dbHandler: DbLogging? = null

    private object Holder {
        val instance: Logger = Logger()
    }

    companion object {
        val instance: Logger by lazy { Holder.instance }
    }

    fun initialize(context: Context) {
        if (dbHandler != null) {
            return
        }
        dbHandler = DbLogging(context)
    }

    fun <T> verbose(@NonNull tag: String, @NonNull description: T) {
        if (loggingEnabled) {
            Log.v(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Verbose)
        }
    }

    fun <T> debug(@NonNull tag: String, @NonNull description: T) {
        if (loggingEnabled) {
            Log.d(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Debug)
        }
    }

    fun <T> info(@NonNull tag: String, @NonNull description: T) {
        if (loggingEnabled) {
            Log.i(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Info)
        }
    }

    fun <T> warning(@NonNull tag: String, @NonNull description: T) {
        if (loggingEnabled) {
            Log.w(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Warning)
        }
    }

    fun <T> error(@NonNull tag: String, @NonNull description: T) {
        if (loggingEnabled) {
            Log.e(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Error)
        }
    }

    private fun <T> tryToWriteToDatabase(@NonNull tag: String, @NonNull description: T, severity: Severity) {
        if (dbHandler != null && writeToDatabaseEnabled) {
            dbHandler?.addLog(
                    DbLog(-1,
                            Date(Calendar.getInstance().timeInMillis),
                            severity,
                            tag,
                            description.toString()))
        }
    }
}